### ISSUE TITLE
IMPORTANT：The definition of the softmax one is wrong

### ISSUE THREAD

**Issue by PhilIp-L-Good** [kyegomez/AttentionIsOFFByOne#5]
The correct definition of softmax one would be:
![image](https://github.com/kyegomez/AttentionIsOFFByOne/assets/89304475/e07e1c3e-07f3-41ff-9a9e-75624463c265)

请参考下面的数学公式来重新考虑：
![image](https://github.com/kyegomez/AttentionIsOFFByOne/assets/89304475/541d2dab-97dc-4399-b1e6-e2fc80253312)

**Comment by PhilIp-L-Good** [comment]
Consider the following mathematical formula to reconsider:

**Comment by PhilIp-L-Good** [comment]
Consider the following mathematical formula to reconsider.

If we multiply both the numerator and denominator of the soft-max function by the constant C, we get the following equation:
![image](https://github.com/kyegomez/AttentionIsOFFByOne/assets/89304475/a98d1ff0-b1d0-4f2a-9c9b-3751dda97407)

**Comment by Devil-SX** [comment]
我不明白您为什么要这么做，可以进一步解释怎么从该数学公式得到将1改为 $e^{-C}$ 的吗

**Comment by martindbp** [comment]
@Devil-SX The reason is that in the original formulation when you subtract the max, it cannot give more than 0.5 attention to the attention sink. But if you replace the 1 with e^c you get:

$\frac{e^{x_i+c}}{e^c + \sum_j e^{x_j+c}} = \frac{e^c e^{x_i}}{e^c + \sum_j e^{x_j}e^c} = \frac{e^c e^{x_i}}{e^c(1 + \sum_j e^{x_j})} = \frac{e^{x_i}}{1 + \sum_j e^{x_j}}$

**Comment by christopher-w-murphy** [comment]
I agree. The definition of softmax_1 is wrong here. I worked with Evan Miller on this, and saw many people make the same mistake. I implemented the correct version in Flash Attention here: https://github.com/softmax1/Flash-Attention-Softmax-N

**Comment by shikhar-srivastava** [comment]
@christopher-w-murphy That looks great! 

Incase anyone needs it, here's a clean implementation: 

```python
def softmax_off_by_one(logits: torch.Tensor, dim: int = -1, dtype: torch.dtype = torch.float32) -> torch.Tensor:
    """Off-by-one softmax (softmax1) with numerical stability.

    softmax1_i(x) = exp(x_i) / (1 + sum_j exp(x_j))
    Stable by shifting by - max(x) along the specified dimension.
    """
    x = logits.to(dtype)
    max_vals = torch.amax(x, dim=dim, keepdim=True)
    shifted = x - max_vals
    exp_shifted = torch.exp(shifted)
    sum_exp = exp_shifted.sum(dim=dim, keepdim=True)
    denom = sum_exp + torch.exp(-max_vals)
    return exp_shifted / denom

```
